### ISSUE TITLE
Excessive space usage

### ISSUE THREAD

**Issue by Luflosi** [dhappy/git-remote-ipfs#3]
I have a git repo where the .git directory uses 27MB according to `du -sh`. I added it to a fresh IPFS datastore using this project and now `du -sh ~/.ipfs` shows that 16GB of space is being used. Is that expected or am I doing something wrong?

**Comment by da2x** [comment]
You’re comparing one specific directory with another that contains can contain all sorts of other things. Remember that it caches things as you browse the web with the IPFS Companion, for instance. Are you sure your IPFS cache doesn’t contain anything else?

Try this: Quit IPFS, move ~/.ipfs to ~/.ipfs-backup, start IPFS again, re-add your repository, and check the ~/.ipfs directory size again.

**Comment by Luflosi** [comment]
Well, I did this in an LXC container where I just installed IPFS and git-remote-igis and didn't use IPFS for anything else. But after doing
```
mv ~/.ipfs ~/.ipfs-backup
ipfs init
ipfs daemon &
cd some/git/repo
git push ipfs:: master
```
the last command returned instantly and the new ~/.ipfs is only a few KB in size. I'll try this again in a completely fresh container.

**Comment by Luflosi** [comment]
I did this again over night. This is exactly what I did:
1. Create a brand new Arch Linux LXC container
2. `pacman -S git gcc go go-ipfs`
3. Switch to normal user
4. `export GOPATH="$HOME/go"`
5. `export GOBIN="$GOPATH/bin"`
6. `mkdir -p "$GOBIN"`
7. `export PATH="$PATH:$GOBIN"`
8. `go get -v github.com/dhappy/git-remote-ipfs/cmd/git-remote-ipfs/...`
9. `cp --no-dereference "$GOPATH/src/github.com/dhappy/git-remote-ipfs/cmd/git-remote-ipfs/git-remote-ipns" "$GOBIN"`
10. `ipfs init`
11. `ipfs daemon &`
12. `git clone https://github.com/kovidgoyal/kitty.git`
13. `cd kitty`
14. `time git push ipfs:: master`
15. Wait seven hours for the command to complete. The command printed text like this: `push: 29442/29443 (P:29441) tree/e296f947e55c38a3a67d379e74baf7a535b8ccf1 (QmUmSkkmJ4tUfJkjCYL21FMztyXojwtgs2W8KbMg)`. The result of `time` is
```
real    439m15.550s
user    0m58.990s
sys     0m36.995s
```
16. `du -sh .git` prints 27MB
17. `du -sh ~/.ipfs` prints 16 GB

**Comment by da2x** [comment]
The repo is 27 MB packed and 198 MB unpacked. IPFS overhead turns those 198 MB into 364 MB, which in turn consumed 432 MB on disk (assuming 4K blocks).

But, yes. I also end up with 16 GB “extra” when following your steps. `ipfs repo gc` cleans it out and still leaves you with a complete and functional repo (but be aware of #5!) Any ideas what the extra data is, @dysbulic?

**Comment by Luflosi** [comment]
Thanks for looking into this.
After pinning the hash and running the gc, `du` now reports 419MB. The IPFS repo is stored on a btrfs filesystem. `compsize` reports that 371MB of disk space would be used if I had btrfs compression turned off (I don't know why there is a difference to `du` 🤷‍♂️) but with compression it only uses 126MB of disk space. I can live with that.
Now we only need to figure out what that extra data is (which btw. was only compressed by 4% by the btrfs compression compared to the 66% after running the gc).

**Comment by dysbulic** [comment]
The difference between reported and deduplicated sizes is definitely a thing.

Each git commit includes a tree that is constructable from the object database. This remote makes each of those trees into an IPFS filesystem, so files are duplicated over and over. This is nearly free in IPFS.

[Temporal pinning service](//temporal.cloud/temporalx/) was the most comprehensible in trying to deal with it, and they introduced a patch that only charged for the deduplicated size, which is pretty significant in these cases.

**Comment by da2x** [comment]
The big question remains, though. What’s with the 19 GB out of nowhere?

> [Temporal] introduced a patch that only charged for the deduplicated size.

Do you have a source for this information? I couldn’t find anything about it on their website or the web at large. I’m the author of the [IPFS pinning services overcharge for duplicated objects](https://www.ctrl.blog/entry/ipfs-pin-storage-accounting.html) article that brought up this problem last year. I discussed it with Temporal at the time, but I haven’t heard anything about them actually addressing it.

**Comment by dysbulic** [comment]
If you go to [play2.temporal.cloud](play2.temporal.cloud) and pin something it will show you the deduplicated size. There was [a Reddit post](https://www.reddit.com/r/ipfs/comments/g3yfh0/temporal_now_supports_deduplicated_pin_size/) about it.

**Comment by dysbulic** [comment]
This could also have been related to the tables I was building of all the commits.

The application has been rewritten in javascript and now uses the same tree structure git does, making updates orders of magnitude faster.

**Comment by Luflosi** [comment]
I did the steps in https://github.com/dhappy/git-remote-ipfs/issues/3#issuecomment-625180977 again but instead of installing via go, I installed git-remote-ipfs via npm. `time git push ipfs:: master` took 720 minutes instead of the 439 minutes before. The size reported via `du -sh .ipfs` is down to 3.5GB from 16GB. After `ipfs repo gc`, `du -sh .ipfs` reports 449MB, roughly the same as before. This is definitely an improvement but there is still quite some space that is taken up for no apparent reason.
